### PR TITLE
Add sensors dynamically based on available device state

### LINKED ISSUE #223: [Architecture] Allow multiple entities per device
**Is your feature request related to a problem? Please describe.**
Currently, we only have one entity per discovered Somfy TaHoma device. However, looking at some of the devices, there are more interesting states and/or actions to take, which we put now in attributes.

Looking at PR's to core, I get the feeling that the attributes part is used less and less and should not be used to store everything. 

For example, looking at some sensors, there are multiple power states and the same for some security devices.

**Describe the solution you'd like**
Rewrite part of the integration to dynamically allow the creation of multiple entities per device.

**Additional context**
Example of a Z-Wave device, with multiple entities. Some less frequent used entities are hidden by default. 
Screenshot 2020-08-12 at 17 44 53

### PR DESCRIPTION
**Pull request by iMicknl**
Fixes #223

Screen Shot 2021-04-18 at 18 31 39

### CODE REVIEW

File: custom_components/tahoma/state_sensor.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    name: str
+    icon: Optional[str] = None
+    unit: Union[None, str, Callable[[dict], str]] = None
+    value: Callable[[Any], Any] = lambda val: val
+    device_class: Optional[str] = None
+    default_enabled: bool = True
+
+
+supported_states = {
+    "core:BatteryState": StateDescription(
```
Comment (tetienne) [inline custom_components/tahoma/state_sensor.py:28]: You don't use the key of this dict.

File: custom_components/tahoma/state_sensor.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        key="core:BatteryState",
```
Comment (tetienne) [inline custom_components/tahoma/state_sensor.py:29]: You can remove the key here and give it to the constructor. It will avoid to duplicate the data.

Comment (iMicknl) [reply custom_components/tahoma/state_sensor.py:29]: Would you know how to pass this?

File: custom_components/tahoma/__init__.py [hunk 3]
```diff
@@ -151,6 +151,7 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
 
     platforms = defaultdict(list)
     platforms[SCENE] = scenarios
+    platforms["sensor"] = []
```
Comment (tetienne) [inline custom_components/tahoma/__init__.py:154]: No constant?

File: custom_components/tahoma/sensor.py [hunk 4]
```diff
@@ -116,6 +117,22 @@ async def async_setup_entry(hass, entry, async_add_entities):
         if device.states
     ]
 
+    for platform, devices in data["platforms"].items():
+        if platform == "scene":
+            continue
+
+        for device in devices:
+            if device.states:
+                for state in device.states:
```
Comment (tetienne) [inline custom_components/tahoma/sensor.py:126]:
```suggestion
            for state in getattr(device, "states", []):
                
```

File: custom_components/tahoma/state_sensor.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    ),
+}
+
+
+class TahomaStateSensor(TahomaEntity, Entity):
+    """Representation of a TaHoma Sensor, based on a secondary device."""
+
+    def __init__(
+        self,
+        device_url: str,
```
Comment (tetienne) [inline custom_components/tahoma/state_sensor.py:108]: I would add the key as a parameter and remove it from the state_description to avoid the duplicate data. https://github.com/iMicknl/ha-tahoma/pull/425/files#r615452018

**Review by tetienne — COMMENTED** [review]
Do you plan to clean the state_attributes method? https://github.com/iMicknl/ha-tahoma/blob/669ac3f1f5448e521cdd2223a6226bb06b319045/custom_components/tahoma/tahoma_entity.py#L80

**Comment by tetienne** [comment]
I really like this PR :)

**Comment by tetienne** [comment]
For my awning:

image

And my Energy meter:

image

For this last, that’s not really readable.

**Comment by iMicknl** [comment]
> And my Energy meter (io:TotalElectricalEnergyConsumptionIOSystemSensor):
> 
> For this last, that’s not really readable.

Lol, I haven't seen a device like this yet. Maybe we should include the device name in the sensor name as well? Is your energy sensor one device, or is this part of our merge based on subsystemid (#1 + #2 etc.)

**Comment by tetienne** [comment]
Yes that’s the most complex devices I own :D That’s indeed a device merging several Somfy device.
```

{JSON
      "creationTime": 1496693986000,
      "lastUpdateTime": 1496693986000,
      "label": "Atlantic",
      "deviceURL": "io://xxxxx#1",
      "shortcut": false,
      "controllableName": "io:TotalElectricalEnergyConsumptionIOSystemSensor",
      "definition": {
        "commands": [],
        "states": [
          {
            "type": "DiscreteState",
            "values": [
              "good",
              "low",
              "normal",
              "verylow"
            ],
            "qualifiedName": "core:DiscreteRSSILevelState"
          },
          {
            "type": "ContinuousState",
            "qualifiedName": "core:ElectricEnergyConsumptionState"
          },
          {
            "type": "ContinuousState",
            "qualifiedName": "core:RSSILevelState"
          },
          {
            "type": "DiscreteState",
            "values": [
              "dead",
              "lowBattery",
              "maintenanceRequired",
              "noDefect"
            ],
            "qualifiedName": "core:SensorDefectState"
          },
          {
            "type": "DiscreteState",
            "values": [
              "available",
              "unavailable"
            ],
            "qualifiedName": "core:StatusState"
          }
        ],
        "dataProperties": [],
        "widgetName": "CumulativeElectricPowerConsumptionSensor",
        "uiProfiles": [
          "ElectricEnergyConsumption"
        ],
        "uiClass": "ElectricitySensor",
        "qualifiedName": "io:TotalElectricalEnergyConsumptionIOSystemSensor",
        "type": "SENSOR"
      },
```

File: custom_components/tahoma/__init__.py (hunk 3 above)
Comment (tetienne) [reply custom_components/tahoma/__init__.py:154]: You can just add the sensors platform manually once we retrieved all the available platforms.

Comment (iMicknl) [reply custom_components/tahoma/__init__.py:154]: > Do you plan to clean the state_attributes method?
> 
> https://github.com/iMicknl/ha-tahoma/blob/669ac3f1f5448e521cdd2223a6226bb06b319045/custom_components/tahoma/tahoma_entity.py#L80

I was thinking about removing this fully eventually, especially since this is not allowed in core. However, in the mean-time it can help us spot attributes / states that we should add.

**Comment by iMicknl** [comment]
> Do you plan to clean the state_attributes method?
> 
> https://github.com/iMicknl/ha-tahoma/blob/669ac3f1f5448e521cdd2223a6226bb06b319045/custom_components/tahoma/tahoma_entity.py#L80

I was thinking about removing this fully eventually, especially since this is not allowed in core. However, in the mean-time it can help us spot attributes / states that we should add.

**Comment by tetienne** [comment]
Do you think you will have some time to finish this PR. It's a nice improvement I would like to have.

**Comment by iMicknl** [comment]
> Do you think you will have some time to finish this PR. It's a nice improvement I would like to have.

I will take some time in the coming weeks, however I could use some help as well :). Feel free to contribute / test as well.

**Comment by tetienne** [comment]
Oki I will have a look too 😊

File: custom_components/tahoma/const.py [hunk 6]
```diff
@@ -76,7 +76,6 @@
     "HitachiAirToAirHeatPump": CLIMATE,  # widgetName, uiClass is HitachiHeatingSystem (not supported)
     "HumiditySensor": SENSOR,
     "Light": LIGHT,
-    "LightSensor": SENSOR,
```
Comment (tetienne) [inline custom_components/tahoma/const.py:79]: TODO: don’t warn when LightSensor detected.

File: custom_components/tahoma/cover_devices/tahoma_cover.py [hunk 7]
```diff
@@ -128,19 +125,6 @@ def is_closed(self):
 
         return None
 
-    @property
-    def icon(self):
```
Comment (tetienne) [inline custom_components/tahoma/cover_devices/tahoma_cover.py:132]: Replace By a sensor.

File: custom_components/tahoma/const.py (hunk 6 above)
Comment (iMicknl) [inline custom_components/tahoma/const.py:79]: @tetienne should we already remove those sensors? Or keep this PR for secondary sensors and in a follow up rewrite our full sensor (and eventually binary sensor) system?

Having our sensor entities different will have other complications if we don't have a 'main' device/entity anymore.

File: custom_components/tahoma/cover_devices/tahoma_cover.py [hunk 8]
```diff
@@ -257,17 +241,6 @@ def is_closing(self):
...
         )
 
-    @property
-    def device_state_attributes(self):
-        """Return the device state attributes."""
-        attr = super().device_state_attributes or {}
-
-        # Obstruction Detected attribute is used by HomeKit
-        if self.has_state(IO_PRIORITY_LOCK_LEVEL_STATE):
-            attr[ATTR_OBSTRUCTION_DETECTED] = True
```
Comment (iMicknl) [inline custom_components/tahoma/cover_devices/tahoma_cover.py:267]: Please don't remove this. Obstruction Detected is used by HomeKit and for now it needs to be a state attribute on the cover.

File: custom_components/tahoma/sensor.py [hunk 9]
```diff
@@ -87,7 +84,7 @@
     "core:TemperatureInCelsius": TEMP_CELSIUS,
     "core:TemperatureInKelvin": TEMP_KELVIN,
     "core:TemperatureInFahrenheit": TEMP_FAHRENHEIT,
-    "core:LuminanceInLux": UNIT_LX,
+    "core:LuminanceInLux": LIGHT_LUX,
```
Comment (iMicknl) [inline custom_components/tahoma/sensor.py:87]: Don't forget to bump the minimum version of HA we need. Maybe we should just remove some other try catch import things soon and require the latest 2021.8 version.

File: custom_components/tahoma/state_sensor.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+
+
+SUPPORTED_STATES = [
+    StateDescription(
+        key="core:BatteryState",
+        name="Battery",
+        unit=PERCENTAGE,
+        device_class=sensor.DEVICE_CLASS_BATTERY,
+    ),
+    StateDescription(
```
Comment (iMicknl) [inline custom_components/tahoma/state_sensor.py:41]: I would love to use `SensorEntityDescription`, see https://github.com/home-assistant/core/pull/53890/files for an example.

We would still need to create our own `OverkizSensorDescription` since we need some additional things like the `value` transformer, but that is fine.

File: custom_components/tahoma/tahoma_entity.py [hunk 11]
```diff
@@ -199,14 +199,6 @@ async def async_cancel_command(self, exec_id: str):
         """Cancel device command in async context."""
         await self.coordinator.client.cancel_command(exec_id)
 
-    def get_base_device_url(self):
```
Comment (iMicknl) [inline custom_components/tahoma/tahoma_entity.py:202]: Why is this removed?

File: custom_components/tahoma/tahoma_entity.py [hunk 12]
```diff
@@ -81,9 +84,6 @@ def device_state_attributes(self) -> Dict[str, Any]:
         """Return the state attributes of the device."""
         attr = {}
 
-        if self.has_state(CORE_RSSI_LEVEL_STATE):
-            attr[ATTR_RSSI_LEVEL] = self.select_state(CORE_RSSI_LEVEL_STATE)
-
         if self.has_state(CORE_BATTERY_STATE):
```
Comment (iMicknl) [inline custom_components/tahoma/tahoma_entity.py:87]: This will be a harder one to migrate? Are there situations where we have two battery states? One with the number and one with just 'normal / low' etc? If that is never the case, we could migrate this one as well.

File: custom_components/tahoma/cover_devices/tahoma_cover.py (hunk 8 above)
Comment (tetienne) [reply custom_components/tahoma/cover_devices/tahoma_cover.py:267]: That’s still the case ?

File: custom_components/tahoma/sensor.py (hunk 9 above)
Comment (tetienne) [reply custom_components/tahoma/sensor.py:87]: Oh I didn’t know it was recently added. In which version it was released ?

File: custom_components/tahoma/cover_devices/tahoma_cover.py (hunk 8 above)
Comment (iMicknl) [reply custom_components/tahoma/cover_devices/tahoma_cover.py:267]: https://github.com/home-assistant/core/blob/236738c455b827413ba052b1ea3bb24214d9f5db/homeassistant/components/homekit/type_covers.py#L177-L182

File: custom_components/tahoma/sensor.py (hunk 9 above)
Comment (iMicknl) [reply custom_components/tahoma/sensor.py:87]: Nevermind, it has been added on Sep 23, 2020... I thought this was also part of a recent commit where they refactored some of these units.

https://github.com/home-assistant/core/commit/6c8e0e20fbd0da03cc5a5d2561a94a4721367bf5

File: custom_components/tahoma/state_sensor.py (hunk 10 above)
Comment (tetienne) [reply custom_components/tahoma/state_sensor.py:41]: Oh clearly that’s something we should do with this PR. I totally missed this new class.

File: custom_components/tahoma/sensor.py [hunk 13]
```diff
@@ -87,7 +84,7 @@
     "core:TemperatureInCelsius": TEMP_CELSIUS,
```
Comment (iMicknl) [inline custom_components/tahoma/sensor.py:84]: This will also make it harder by the way. If we move all sensor entities to this new structure, we cannot use this list easily. `"core:MeasuredValueType"` does only exist once for a device, where it can have multiple states. We can only use this when we know what the primary state is of a certain ui class/widget.

File: custom_components/tahoma/tahoma_entity.py (hunk 11 above)
Comment (tetienne) [reply custom_components/tahoma/tahoma_entity.py:202]: I didn’t know what name gave to my function as I want to return the base url and the index. :S

Comment (iMicknl) [reply custom_components/tahoma/tahoma_entity.py:202]: `split_device_url`? Don't know either..

File: custom_components/tahoma/state_sensor.py (hunk 10 above)
Comment (iMicknl) [reply custom_components/tahoma/state_sensor.py]: Done.

File: custom_components/tahoma/const.py (hunk 6 above)
Comment (tetienne) [reply custom_components/tahoma/const.py:79]: So should I remove also my StateDescription related to the light?

File: custom_components/tahoma/tahoma_entity.py (hunk 11 above)
Comment (tetienne) [reply custom_components/tahoma/tahoma_entity.py:202]: Ahah you see :D

File: custom_components/tahoma/tahoma_entity.py [hunk 14]
```diff
@@ -3,42 +3,20 @@
...
-ATTR_RSSI_LEVEL = "rssi_level"
-
-CORE_AVAILABILITY_STATE = "core:AvailabilityState"
-CORE_BATTERY_STATE = "core:BatteryState"
 CORE_MANUFACTURER = "core:Manufacturer"
 CORE_MANUFACTURER_NAME_STATE = "core:ManufacturerNameState"
 CORE_MODEL_STATE = "core:ModelState"
 CORE_PRODUCT_MODEL_NAME_STATE = "core:ProductModelNameState"
-CORE_RSSI_LEVEL_STATE = "core:RSSILevelState"
-CORE_SENSOR_DEFECT_STATE = "core:SensorDefectState"
```
Comment (tetienne) [inline custom_components/tahoma/tahoma_entity.py:23]: We need this one. I just checked, and it is used when there is no more battery. For instance:

image

```JSON
{
	"states": [
		{
			"name": "core:StatusState",
			"type": 3,
			"value": "unavailable"
		},
		{
			"name": "core:SensorDefectState",
			"type": 3,
			"value": "dead"
		},
		{
			"name": "core:DiscreteRSSILevelState",
			"type": 3,
			"value": "verylow"
		},
		{
			"name": "core:RSSILevelState",
			"type": 2,
			"value": 4
		},
		{
			"name": "core:LuminanceState",
			"type": 2,
			"value": 110
		}
	]
}
```

Or perhaps do you plan to implement a battery sensor ?

File: custom_components/tahoma/cover_devices/tahoma_cover.py (hunk 8 above)
Comment (tetienne) [reply custom_components/tahoma/cover_devices/tahoma_cover.py]: Done.

File: custom_components/tahoma/cover_devices/tahoma_cover.py [hunk 15]
```diff
@@ -64,6 +64,8 @@
 ICON_LOCK_ALERT = "mdi:lock-alert"
```
Comment (iMicknl) [inline custom_components/tahoma/cover_devices/tahoma_cover.py]: Am I correct that we don't use these icons anymore?

File: custom_components/tahoma/tahoma_entity.py (hunk 14 above)
Comment (iMicknl) [reply custom_components/tahoma/tahoma_entity.py:23]: Good to know! I didn't know that this was an important state. The issue is that battery state was coming from multiple properties...

What would you suggest? Add them as multiple sensors or create a separate battery sensor entity? My preference is to have multiple sensors, if the three states are not available at the same time in a device. Than 3 different sensors make sense. However I don't want to add 2 battery sensors accidentally for a device.

Comment (tetienne) [reply custom_components/tahoma/tahoma_entity.py:23]: I think a battery sensor will make more sense, and reduce code complexity. And it will avoid as you said to have multiple sensors linked to the same topic.

Comment (iMicknl) [reply custom_components/tahoma/tahoma_entity.py:23]: Agree! However, are we sure that the battery state can exist multiple times?

File: custom_components/tahoma/cover_devices/tahoma_cover.py (hunk 7 above)
Comment (tetienne) [reply custom_components/tahoma/cover_devices/tahoma_cover.py:132]: @iMicknl About this one, I just checked, the state is returned only when the issue occured. Meaning we cannot create the sensor during the setup 99% of the time. What do you think about hardcoding it ?

Comment (iMicknl) [reply custom_components/tahoma/cover_devices/tahoma_cover.py:132]: @tetienne hmm, this is a hard one. Should we create an issue for this one and tackle it in a follow up PR? 
I think that `CORE_PRIORITY_LOCK_TIMER_STATE` is always present, where `IO_PRIORITY_LOCK_ORIGINATOR_STATE` is only present when there is a reason (lock timer state > 0).

File: custom_components/tahoma/tahoma_entity.py (hunk 14 above)
Comment (iMicknl) [reply custom_components/tahoma/tahoma_entity.py:23]: @tetienne should we create an issue for this and do it in a follow up PR? Let's get this huge PR merged asap.

Comment (tetienne) [reply custom_components/tahoma/tahoma_entity.py:23]: Yep, I agree. Just roll back the battery code to be sure we still have the battery logic somewhere.

File: custom_components/tahoma/const.py (hunk 6 above)
Comment (tetienne) [reply custom_components/tahoma/const.py:79]: At the end, we should ignore all the SENSOR devices. But yes, for the moment we can add it the ignore list.

File: custom_components/tahoma/tahoma_entity.py (hunk 14 above)
Comment (iMicknl) [reply custom_components/tahoma/tahoma_entity.py:23]: Done.

File: custom_components/tahoma/const.py [hunk 16]
```diff
@@ -39,13 +38,26 @@
...
+    "SunIntensitySensor",
+    "SunSensor",
+    "GasSensor",
+    "GenericSensor",
+    "HumiditySensor",
+    "TemperatureSensor",
+    "ThermalEnergySensor",
+    "WaterSensor",
+    "WeatherSensor",
+    "WindSensor",
```
Comment (tetienne) [inline custom_components/tahoma/const.py:54]: You can add LightSensor.

File: custom_components/tahoma/cover_devices/tahoma_cover.py (hunk 7 above)
Comment (tetienne) [reply custom_components/tahoma/cover_devices/tahoma_cover.py:132]: Agreed. We can rollback this and focus on it later.

File: custom_components/tahoma/sensor.py [hunk 17]
```diff
@@ -1,116 +1,338 @@
...
+    OverkizSensorDescription(
+        key="core:BatteryState",
+        name="Battery",
+        device_class=sensor.DEVICE_CLASS_BATTERY,
+        value=lambda value: str(value).capitalize(),
+    ),
+    OverkizSensorDescription(
+        key="core:RSSILevelState",
+        name="RSSI Level",
+        value=lambda value: round(value),
```
Comment (tetienne) [inline custom_components/tahoma/sensor.py:61]: Why round the value for every sensor? The raw data is not correct?

File: custom_components/tahoma/sensor.py [hunk 18]
```diff
@@ -1,116 +1,338 @@
...
+        unit_of_measurement=SIGNAL_STRENGTH_DECIBELS,
+        device_class=sensor.DEVICE_CLASS_SIGNAL_STRENGTH,
+        state_class=STATE_CLASS_MEASUREMENT,
+    ),
+    OverkizSensorDescription(
+        key="core:ExpectedNumberOfShowerState",
```
Comment (tetienne) [inline custom_components/tahoma/sensor.py:67]: This one is magic!

File: custom_components/tahoma/sensor.py [hunk 19]
```diff
@@ -1,116 +1,338 @@
...
+        key="io:InletEngineState",
+        name="Inlet Engine",
+        icon="mdi:fan-chevron-up",
+        value=lambda value: round(value),
+        unit_of_measurement=VOLUME_FLOW_RATE_CUBIC_METERS_PER_HOUR,
+        state_class=STATE_CLASS_MEASUREMENT,
+    ),
+    OverkizSensorDescription(
+        key="hlrrwifi:RoomTemperatureState",
+        name="Room Temperature",
```
Comment (tetienne) [inline custom_components/tahoma/sensor.py:116]: To ease the code, we can generate the name from the key.

File: custom_components/tahoma/sensor.py [hunk 20]
```diff
@@ -1,116 +1,338 @@
...
+        icon="mdi:alert",
+    ),
+    OverkizSensorDescription(
+        key="core:FossilEnergyConsumptionState",
+        name="Fossil Energy Consumption",
+        device_class=sensor.DEVICE_CLASS_ENERGY,
+    ),
+    OverkizSensorDescription(
+        key="core:GasConsumptionState",
+        name="Gas Consumption",
```
Comment (tetienne) [inline custom_components/tahoma/sensor.py:140]: No device class?

File: custom_components/tahoma/sensor.py [hunk 21]
```diff
@@ -122,68 +344,59 @@ async def async_setup_entry(
...
-            CORE_LUMINANCE_STATE,
-            CORE_RELATIVE_HUMIDITY_STATE,
-            CORE_SUN_ENERGY_STATE,
-            CORE_TEMPERATURE_STATE,
-            CORE_THERMAL_ENERGY_CONSUMPTION_STATE,
-            CORE_WINDSPEED_STATE,
-            CORE_WATER_CONSUMPTION_STATE,
-        )
-        return round(state, 2) if state is not None else None
+        state = self.device.states[self.entity_description.key]
```
Comment (tetienne) [inline custom_components/tahoma/sensor.py:384]: We can do the same for the binary sensor pr.

File: custom_components/tahoma/sensor.py (hunk 20 above)
Comment (iMicknl) [reply custom_components/tahoma/sensor.py:140]: afaik, there is no device class for gas consumption sensors.

File: custom_components/tahoma/sensor.py (hunk 19 above)
Comment (iMicknl) [reply custom_components/tahoma/sensor.py:116]: I thought about this as well, however I don't see the value. This lookup is fast and gives us a way to easily change the name.

We can always add this in a follow up PR? However, I think there are more pressing issues to solve :-).

File: custom_components/tahoma/sensor.py (hunk 17 above)
Comment (iMicknl) [reply custom_components/tahoma/sensor.py:61]: Tbh, I don't recall why I added this. There are quite some states that give very precise measures, thus I chose to round it. 

Should I remove all the rounds and add them back based on user reports? Or should we keep it safe and just keep them?

File: custom_components/tahoma/const.py (hunk 16 above)
Comment (iMicknl) [reply custom_components/tahoma/const.py:55]: Done

File: custom_components/tahoma/cover_devices/tahoma_cover.py (hunk 7 above)
Comment (iMicknl) [reply custom_components/tahoma/cover_devices/tahoma_cover.py:132]: Done

File: custom_components/tahoma/sensor.py (hunk 19 above)
Comment (tetienne) [reply custom_components/tahoma/sensor.py:116]: Sure. It’s just I like to reduce number of lines :D

File: custom_components/tahoma/sensor.py (hunk 17 above)
Comment (tetienne) [reply custom_components/tahoma/sensor.py:61]: I would suggest to keep the raw value. And see if user complains.

**Review by tetienne — DISMISSED** [review]
Tested locally. Seems to be OK. I can even now add my energy sensors within the Energy panel.

File: custom_components/tahoma/sensor.py (hunk 17 above)
Comment (iMicknl) [reply custom_components/tahoma/sensor.py]: Done! I just added it for RSSI, since I can test this one and it always return .0, thus rounding will make it cleaner.

**Review by tetienne — APPROVED** [review]
